sklearn/metrics/pairwise.py

```diff
@@ -375,24 +380,30 @@ def _euclidean_distances(X, Y, X_norm_squared=None, Y_norm_squared=None, squared
     float32, norms needs to be recomputed on upcast chunks.
     TODO: use a float64 accumulator in row_norms to avoid the latter.
     """
-    if X_norm_squared is not None and X_norm_squared.dtype != np.float32:
-        XX = X_norm_squared.reshape(-1, 1)
-    elif X.dtype != np.float32:
-        XX = row_norms(X, squared=True)[:, np.newaxis]
+    xp, _, device_ = get_namespace_and_device(X, Y)
+    if X_norm_squared is not None and X_norm_squared.dtype != xp.float32:
+        XX = xp.reshape(X_norm_squared, (-1, 1))
+    elif X.dtype != xp.float32:
+        XX = row_norms(X, squared=True)[:, None]
     else:
         XX = None
 
     if Y is X:
         YY = None if XX is None else XX.T
     else:
-        if Y_norm_squared is not None and Y_norm_squared.dtype != np.float32:
-            YY = Y_norm_squared.reshape(1, -1)
-        elif Y.dtype != np.float32:
-            YY = row_norms(Y, squared=True)[np.newaxis, :]
+        if Y_norm_squared is not None and Y_norm_squared.dtype != xp.float32:
+            YY = xp.reshape(Y_norm_squared, (1, -1))
+        elif Y.dtype != xp.float32:
+            YY = row_norms(Y, squared=True)[None, :]
         else:
             YY = None
 
-    if X.dtype == np.float32 or Y.dtype == np.float32:
+    if _max_precision_float_dtype(xp=xp, device=device_) == xp.float32:
+        # special case for mps devices which don't support float64.
+        X_r = X[:, None]
+        Y_r = Y[None, :]
+        distances = xp.sum((X_r - Y_r) ** 2, axis=2)
```

Comment:
why is this code different than the case where the data is already `float64`?

Reply:
I think this might be more numerically stable when we consider float32. The default code for float64 computes the distance by breaking it up into three parts.

Reply:
It's more numerically stable but I am afraid it might blow-up memory usage to store the intermediate `(X[:, None] - Y[None, :]) ** 2` array. We would probably need to chunk this computation in a loop similarly to what we do in `_euclidean_distances_upcast`. However this will make this code much more complex.

I think we might want to do this work in a dedicated PR and leave with the fact that `euclidean_distances` will be numerically unstable on the MPS device for the time being (as long as it's a documented limitation in our array API doc).

Reply:
I have reverted this to the previous state.